### ISSUE TITLE
RMWC Typescript Integration with MDC 1.0.0 prerelease issues

### ISSUE THREAD

**Issue by jamesmfriedman** [material-components/material-components-web#4463]
There are likely to be a litany of issues that pop up from going through this since RMWC is in Typescript, and also now uses the foundation / adapter pattern.

- [x] Lerna bootstrap issue (caused by find and replace issue on my end)
- [x] `mdc-*` Module declaration causes errors (fixed by mat-web correcting module names in the definition files)
- [x] Definition issues with extending MDCFoundation and redeclaring `static attachTo`
- [x] Definition issues with adapter_ being protected on MDCFoundation on classes that are trying to extend it
- [x] FloatingLabel.getWidth adapter should accept null as result in the event the DOM element is missing.
- [x] MenuSurface.getInnerDimensions should allow either null as a result, or allow width and height to be null.

**Comment by jamesmfriedman** [comment]
The above may be user error ^. I'll double check in a bit, but this may be due to a find a replace all updating the package.lock file. I'll make sure I clear those out and try again.

**Comment by jamesmfriedman** [comment]
It seems like the module declarations are incorrect, or I'm consuming them wrong?

All module declarations declared as `mdc-*` break.
<img width="1680" alt="screen shot 2019-03-02 at 11 17 19 am" src="https://user-images.githubusercontent.com/975256/53684648-c7db7c80-3cdd-11e9-9956-b33a3fe3de42.png">

Changing the declaration to its actual import path fixes the problem, in this case `@material/checkbox`.'' Ignore the fact that there are additional errors for now ;)
<img width="1680" alt="screen shot 2019-03-02 at 11 18 28 am" src="https://user-images.githubusercontent.com/975256/53684665-1426bc80-3cde-11e9-87af-c74419b4f209.png">

**Comment by kfranqueiro** [comment]
The first error encountered in James' last post is due to us incorrectly naming the modules we generate using dts-bundle. The second error is legitimate, due to an adapter API that no longer exists.

**Comment by jamesmfriedman** [comment]
Awesome. I’ll look for updates on this for a new version.

**Comment by jamesmfriedman** [comment]
Issue extending class and overriding `static attachTo`

![image](https://user-images.githubusercontent.com/975256/53811970-7a068480-3f28-11e9-8b77-8f2e4230aaa3.png)

Potential Fix: Don't declare attachTo on MDCComponent

**Comment by jamesmfriedman** [comment]
Issue passing generic constraint to MDCComponent. This seems to result from the fact that `adapter_` is protected in MDComponent which causes issues with extending it.

![image](https://user-images.githubusercontent.com/975256/53812087-b89c3f00-3f28-11e9-88c4-4d594860b240.png)

Potential Fix: Making `MDCComponent.adapter_` public resolves this issue.

**Comment by jamesmfriedman** [comment]
Issue with FloatingLabel.getWidth requiring a number. In React, we can't guarantee we have a DOMNode to actually get the width from.

![image](https://user-images.githubusercontent.com/975256/53812395-60b20800-3f29-11e9-960a-a5d4b17cfc86.png)

Potential Fix: Allow null and handle in the foundation, or let me know if passing `0` is a safe alternative.

**Comment by kfranqueiro** [comment]
RE FloatingLabel, returning 0 from getWidth is probably safe (it's what our placeholder in the default adapter does, and you're clearly not going to be doing anything else with that value if it's not in the DOM anyway).

RE the other two issues, we've been having trouble reproducing these errors in the MDC Web codebase or a vanilla example, though at the very least the `attachTo` one is something I'm surprised we don't see on our end. I'm a little bit baffled by the other error though, unless it's complaining because MDCCheckboxFoundation relies on a different adapter...

**Comment by jamesmfriedman** [comment]
![image](https://user-images.githubusercontent.com/975256/53854019-d94eae00-3f95-11e9-917e-cbc55e7cdc21.png)

Woot woot! Got a green build (I've tuned the above warnings out for a while, this is a green build for me). A quick debriefing on this one:

- Doing this with Typescript makes the integration way more sane. I was able to fix a bunch of things that had probably been bugs for a long long time.
- The issue with the module namespace in UMD bundles was a real thing, thanks for the fix on that.
- Issues where the adapters required a number for things like offsetWidth were plentiful. I took the advice above and just used a value of zero whenever I didn't have a domNode to compute something from.
- I still believe the issues that are showing up above in the MDCComponent definitions are real issues. It took me forever to track down what was causing the issue...

A  quick debrief on how to reproduce it (in RMWC at least)
- RMWC uses lerna like mat-web.
- running `npx lerna bootstrap` did its job like it was supposed to and linked all the deps
- for some unknown reason, when  running in this mode specifically, I get the type error. I looked into those specific errors on Google / Stack Overflow / Github and it appears they are hard to troubleshoot anytime they pop up.
- My fix was to change the bootstrap command to `npx lerna bootstrap --hoist`. Whenever the packages are hoisted, the issue is not present. Your guess is as good as mine as to why, but it likely has to do with some deep issue between typescript module resolution, node module resolution, and symlinking. I pulled every single thing out until it was just a vanilla Create React App build and the issue was still present, so this wasn't caused by some weird configuration I might have had.
- It is HIGHLY unlikely anyone else will ever encounter this, but when I was digging through the actual type issues, they do seem like they are actual type issues, but for whatever reason they're not being triggered for other people.
- One more thing to note, pointing any of the @material packages package.json `types` key at the non umd build `d.ts` file corrected the issue as well, so the issue is specific to the UMD bundles definitions. 🤦‍♂️

Long story short, Javascript is weird. Feel free to close this issue unless anyone thinks the above (potential) type issues should be fixed.

**Comment by kfranqueiro** [comment]
This is definitely baffling... I'm equally baffled as to how we _haven't_ reproduced it within MDC Web. I do agree that e.g. the attachTo error seems legitimate, but unless we encounter it within the context of MDC Web (so that we can verify before/after fixing) I'm inclined to close this and not block on it. I'm glad you were able to get RMWC working!
